hcl2template/types.packer_config.go

```diff
@@ -284,12 +285,52 @@ func (c *PackerConfig) evaluateLocalVariable(local *LocalBlock) hcl.Diagnostics
 func (cfg *PackerConfig) evaluateDatasources(skipExecution bool) hcl.Diagnostics {
 	var diags hcl.Diagnostics
 
+	dependencies := map[DatasourceRef][]DatasourceRef{}
 	for ref, ds := range cfg.Datasources {
 		if ds.value != (cty.Value{}) {
 			continue
 		}
+		// Pre-examine body of this data source to see if it uses another data
+		// source in any of its input expressions. If so, skip evaluating it for
+		// now, and add it to a list of datasources to evaluate again, later,
+		// with the datasources in its context.
+		// This is essentially creating a very primitive DAG just for data
+		// source interdependencies.
+		block := ds.block
+		body := block.Body
+		attrs, _ := body.JustAttributes()
+
+		skipFirstEval := false
+		for _, attr := range attrs {
+			vars := attr.Expr.Variables()
+			for _, v := range vars {
+				// check whether the variable is a data source
+				if v.RootName() == "data" {
+					// construct, backwards, the data source type and name we
+					// need to evaluate before this one can be evaluated.
+					dependsOn := DatasourceRef{
+						Type: v[1].(hcl.TraverseAttr).Name,
+						Name: v[2].(hcl.TraverseAttr).Name,
+					}
+					log.Printf("The data source %#v depends on datasource %#v", ref, dependsOn)
+					if dependencies[ref] != nil {
+						dependencies[ref] = append(dependencies[ref], dependsOn)
+					} else {
+						dependencies[ref] = []DatasourceRef{dependsOn}
+					}
+					skipFirstEval = true
+				}
+			}
+		}
```

Comment:
I think this could probably be a function, where we pass a HCL block and that gives us a full list of dependencies, recursively. I precises recursively here because if a reference is being made in a nested block, this will ignore it; since we are calling `JustAttributes`.
Furthermore, I think that this should probably be done in the end or after `decodeDataBlock`; and that we should store all 'dependencies' in `DatasourceBlock`.

Reply:
We discussed this offline -- since this case doesn't actually exist right now, I'll come back to this comment in its own PR so we can get this merged and ready for customers.